pkg/domain/entities/params/generator/generator.go

```diff
@@ -0,0 +1,174 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"text/template"
+)
+
+var bodyTmpl = `package {{.PackageName}}
+
+import (
+{{range $import := .Imports}}	{{$import}}
+{{end}}
+)
+`
+
+var fieldTmpl = `
+// With{{.Name}}
+func(o *{{.StructName}}) With{{.Name}}(value {{.Type}}) *{{.StructName}} {
+	v := &value
+	o.{{.Name}} = v
+	return o
+}
+`
+
+var changed = `
+// Changed
+func (o *{{.StructName}}) Changed(fieldName string) bool {
+	r := reflect.ValueOf(o)
+	value := reflect.Indirect(r).FieldByName(fieldName)
+	return !value.IsNil()
+}
+`
+
+func writeChangedFunc(f *os.File, structName string) error {
+	changedStruct := struct {
+		StructName string
+	}{
+		StructName: structName,
+	}
+
+	t, err := template.New(structName).Parse(changed)
+	if err != nil {
+		return err
+	}
+	return t.Execute(f, changedStruct)
+}
+
+func main() {
+	var closed bool
+	srcFile := os.Getenv("GOFILE")
+	pkg := os.Getenv("GOPACKAGE")
+	inputStructName := os.Args[1]
+	b, err := ioutil.ReadFile(srcFile)
+	if err != nil {
+		panic(err)
+	}
+	fset := token.NewFileSet() // positions are relative to fset
+	f, err := parser.ParseFile(fset, "", b, parser.ParseComments)
+	if err != nil {
+		panic(err)
+	}
+
+	// always add reflect
+	imports := []string{"\"reflect\""}
+	for _, imp := range f.Imports {
+		imports = append(imports, imp.Path.Value)
+	}
+
+	out, err := os.Create(strings.ToLower(inputStructName) + "_" + srcFile)
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		if !closed {
+			out.Close()
+		}
+	}()
+	bodyStruct := struct {
+		PackageName string
+		Imports     []string
+	}{
+		PackageName: pkg,
+		Imports:     imports,
+	}
+	bTmpl, err := template.New(inputStructName).Parse(bodyTmpl)
+	if err != nil {
+		panic(err)
+	}
+	if err := bTmpl.Execute(out, bodyStruct); err != nil {
+		panic(err)
+	}
+
+	var structNode ast.Node
+	ast.Inspect(f, func(n ast.Node) bool {
+		ref, refOK := n.(*ast.TypeSpec)
+		if refOK {
+			if ref.Name.Name == inputStructName {
+				structNode = n
+				if err := writeChangedFunc(out, inputStructName); err != nil {
+					panic(err)
+				}
+
+				x := ref.Type.(*ast.StructType)
+				for _, field := range x.Fields.List {
+					var (
+						name string
+					)
+					typeExpr := field.Type
+					start := typeExpr.Pos() - 1
+					end := typeExpr.End() - 1
+					fieldType := strings.Replace(string(b[start:end]), "*", "", 1)
+					if len(field.Names) > 0 {
+						name = field.Names[0].Name
+						if len(name) < 1 {
+							panic(errors.New("bad name"))
+						}
+					}
+					tmplStruct := struct {
+						Name       string
+						StructName string
+						Type       string
+					}{
+						Name:       name,
+						StructName: inputStructName,
+						Type:       fieldType,
+					}
+					t, err := template.New(name).Parse(fieldTmpl)
```

Comment:
I would move this parsing out of the loop. Again no need to parse and store a template per field.

Should the field tags be copied over as well from the source?

Reply:
field tags?